pkg/executor/aggregate/agg_hash_final_worker.go

```diff
@@ -73,15 +95,31 @@ func (w *HashAggFinalWorker) initBInMap() {
 	}
 }
 
-func (w *HashAggFinalWorker) consumeIntermData(sctx sessionctx.Context) (err error) {
-	for {
+func (w *HashAggFinalWorker) consumeIntermData(sctx sessionctx.Context) (spillContinue bool, err error) {
```

Comment:
``` suggestion
func (w *HashAggFinalWorker) consumeIntermData(sctx sessionctx.Context) (spillContinue bool, err error) {
    defer func() { w.isFirstInput = true }()
    var input *aggfuncs.AggPartialResultMapper
    var ok bool
    continueLoop := true
    for continueLoop {
        waitStart := time.Now()
        input, ok, continueLoop = w.getInputData()
		if w.stats != nil {
			w.stats.WaitTime += int64(time.Since(waitStart))
		}
        if !ok {
            return false, nil
        }

        if w.isFirstInput {
            w.handleFirstInput(input)
            continue
        }

        failpoint.Inject("ConsumeRandomPanic", nil)

        if err := w.mergeInputIntoResultMap(sctx, input); err != nil {
            return false, err
        }
    }
    return true, nil
}

func (w *HashAggFinalWorker) getInputData() (*aggfuncs.AggPartialResultMapper, bool, bool) {
    if w.isSpilledTriggered {
        return w.getInputFromDisk()
    }
    input, ok := w.getPartialInput()
    return input, ok, true
}

func (w *HashAggFinalWorker) handleFirstInput(input *aggfuncs.AggPartialResultMapper) {
    w.isFirstInput = false
    w.partialResultMap = *input
    w.initBInMap()
}

func (w *HashAggFinalWorker) mergeInputIntoResultMap(sctx sessionctx.Context, input *aggfuncs.AggPartialResultMapper) error {
    execStart := time.Now()
    allMemDelta := int64(0)
    for key, value := range *input {
        dstVal, ok := w.partialResultMap[key]
        if !ok {
            w.handleNewGroupKey(key, value)
            continue
        }

        for j, af := range w.aggFuncs {
            memDelta, err := af.MergePartialResult(sctx, value[j], dstVal[j])
            if err != nil {
                return err
            }
            allMemDelta += memDelta
        }
    }
    w.memTracker.Consume(allMemDelta)
	if w.stats != nil {
		w.stats.ExecTime += int64(time.Since(execStart))
		w.stats.TaskNum++
	}
    return nil
}

func (w *HashAggFinalWorker) handleNewGroupKey(key string, value []aggfuncs.PartialResult) {
    if len(w.partialResultMap)+1 > (1<<w.BInMap)*hack.LoadFactorNum/hack.LoadFactorDen {
        w.memTracker.Consume(hack.DefBucketMemoryUsageForMapStrToSlice * (1 << w.BInMap))
        w.BInMap++
    }
    w.partialResultMap[key] = value
}
```

Reply:
done